### PR TITLE
Added support for placing redstone dust

### PR DESCRIPTION
**Pull request by ToastyMallows**

### CODE REVIEW

File: TrueCraft.Core/Logic/Items/RedstoneItem.cs [hunk 1]
```diff
@@ -1,14 +1,50 @@
...
+using TrueCraft.Core.Logic.Blocks;
 
 namespace TrueCraft.Core.Logic.Items
 {
     public class RedstoneItem : ItemProvider
     {
+        /// <summary>
+        /// Redstone cannot be placed on any of the following BlockIDs
+        /// </summary>
+        protected static byte[] cannotPlace = { RedstoneDustBlock.BlockID, GlassBlock.BlockID, AirBlock.BlockID };
```
Comment (SirCmpwn) [inline TrueCraft.Core/Logic/Items/RedstoneItem.cs]: Don't do this - instead get the block provider and check its opacity.

File: TrueCraft.Core/Logic/Items/RedstoneItem.cs [hunk 2]
```diff
@@ -1,14 +1,50 @@
...
+
+        public override void ItemUsedOnBlock(Coordinates3D coordinates, ItemStack item, BlockFace face, IWorld world, IRemoteClient user)
+        {
+            // TODO: Running backwards and spamming redstone dust sometimes causes redstone to be placed on redstone
+
+            if (face != BlockFace.PositiveY)
+            {
+                // Redstone dust cannot be placed anywhere but the top of a block
+                return;
+            }
```
Comment (SirCmpwn) [inline TrueCraft.Core/Logic/Items/RedstoneItem.cs]: Clicking the side of the pink block places the redstone as shown.

**Comment by SirCmpwn** [comment]
The placement behavior is just like a block, there's no incorrect faces of a neighboring block to click. It just has to check if the block it would end up is valid (i.e. there is an opaque block underneath it).

File: TrueCraft.Core/Logic/Items/RedstoneItem.cs (hunk 1 above)
Comment (ToastyMallows) [inline TrueCraft.Core/Logic/Items/RedstoneItem.cs]: Redstone Dust block has its Opacity set to false, is this right?

File: TrueCraft.Core/Logic/Items/RedstoneItem.cs (hunk 2 above)
Comment (ToastyMallows) [inline TrueCraft.Core/Logic/Items/RedstoneItem.cs]: Is this expected behavior?

File: TrueCraft.Core/Logic/Items/RedstoneItem.cs (hunk 1 above)
Comment (SirCmpwn) [inline TrueCraft.Core/Logic/Items/RedstoneItem.cs]: Yes, that's correct.

File: TrueCraft.Core/Logic/Items/RedstoneItem.cs (hunk 2 above)
Comment (SirCmpwn) [inline TrueCraft.Core/Logic/Items/RedstoneItem.cs]: What?

Comment (ToastyMallows) [inline TrueCraft.Core/Logic/Items/RedstoneItem.cs]: Is this what should happen?  When you click on the side of the pink block should redstone dust be placed on the white block?

Comment (robinkanters) [inline TrueCraft.Core/Logic/Items/RedstoneItem.cs]: I'd expect it to, yes

Comment (SirCmpwn) [inline TrueCraft.Core/Logic/Items/RedstoneItem.cs]: Yes. It should act like placing any other block, but with a requirement for opaque support in the down direction.

**Comment by ToastyMallows** [comment]
Something like this?

```
coordinates += MathHelper.BlockFaceToCoordinates(face);
IBlockProvider supportingBlock = world.BlockRepository.GetBlockProvider(world.GetBlockID(coordinates));

if (!supportingBlock.Opaque)
{
    world.SetBlockID(coordinates, RedstoneDustBlock.BlockID);
    item.Count--;
    user.Inventory[user.SelectedSlot] = item;
}
```

**Comment by SirCmpwn** [comment]
No, the supported block would be underneath the redstone, not at those coordinates. Get the supporting block with `coordinates + Coordinates3D.Down`.

**Comment by ToastyMallows** [comment]
Ok now I have this but its leading to unexpected behavior, when I click on the same block a redstone dust block is placed underneath the one I just placed.

```
IBlockProvider supportingBlock = world.BlockRepository.GetBlockProvider(world.GetBlockID(coordinates += Coordinates3D.Down));

if (supportingBlock.Opaque)
{
    world.SetBlockID(coordinates, RedstoneDustBlock.BlockID);
    item.Count--;
    user.Inventory[user.SelectedSlot] = item;
}
```

**Comment by ToastyMallows** [comment]
Oops I think I see the problem

**Comment by ToastyMallows** [comment]
@SirCmpwn I think I figured it out

**Comment by SirCmpwn** [comment]
Also wire it up to `BlockProvider`'s support code (i.e. what torches use) so that when you destroy the block underneath the redstone, the redstone is destroyed.

**Comment by ToastyMallows** [comment]
@SirCmpwn Gah I didn't mean to put the upstream merge commit into this.  Hopefully you can just remove that?

**Comment by SirCmpwn** [comment]
It doesn't really matter.

**Pull request merged**